File: tests/python/greedy_algorithms/task_assignment/test_task_assignment.py
from problems.greedy_algorithms.task_assignment.python.task_assignment import \
    task_assignment
from tests.python.utility import load_test_cases


def convert_to_set(pairs: list[list[int]]) -> set[frozenset[int]]:
    result = set()
    for x, y in pairs:
        result.add(frozenset([x, y]))

    return result


def is_valid_result(result: list[list[int]], expected: list[list[int]]) -> bool:
    expected_set = convert_to_set(expected)

    for x, y in result:
        direct_pair = frozenset([x, y])

        if direct_pair not in expected_set:
            partial_match = any(x in pair or y in pair for pair in expected_set)
            if not partial_match:
                return False

    return True


def test_task_assignment() -> None:
    test_cases = load_test_cases("greedy_algorithms/task_assignment.json")

    for idx, test in enumerate(test_cases):
        k = test["k"]
        tasks = test["tasks"]
        expected = test["expected"]

        result = task_assignment(k, tasks)

        is_valid = is_valid_result(result, expected)

        assert (
            is_valid == True
        ), f"Test case {idx} failed: expected {expected}, got {result}"

File: tests/cpp/greedy_algorithms/test_task_assignment.cpp
#include "../../../problems/greedy_algorithms/task_assignment/cpp/task_assignment.h"
#include "../jsontestbase.h"

#include <set>
#include <vector>

class TaskAssignmentTest : public JsonTestBase {
public:
  TaskAssignmentTest() {
    json_file_path =
        std::string(TEST_CASES_DIR) + "/greedy_algorithms/task_assignment.json";
  }
};

std::set<std::set<int>>
convertToSet(const std::vector<std::vector<int>> &pairs) {
  std::set<std::set<int>> result;
  for (const auto &pair : pairs) {
    result.insert({pair[0], pair[1]});
  }
  return result;
}

bool isValidResult(const std::vector<std::vector<int>> &result,
                   const std::vector<std::vector<int>> &expected) {
  auto expectedSet = convertToSet(expected);

  for (const auto &pair : result) {
    std::set<int> directPair = {pair[0], pair[1]};
    if (expectedSet.find(directPair) == expectedSet.end()) {
      bool partialMatch = false;
      for (const auto &expectedPair : expectedSet) {
        if (expectedPair.count(pair[0]) > 0 ||
            expectedPair.count(pair[1]) > 0) {
          partialMatch = true;
          break;
        }
      }
      if (!partialMatch)
        return false;
    }
  }
  return true;
}

TEST_F(TaskAssignmentTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty";

  for (const auto &test : test_cases) {
    auto k = test["k"].get<int>();
    auto tasks = test["tasks"].get<std::vector<int>>();
    auto expected = test["expected"].get<std::vector<std::vector<int>>>();

    auto result = taskAssignment(k, tasks);

    EXPECT_TRUE(isValidResult(result, expected))
        << "Failed for test case: " << test.dump(2);
  }
}
